kore/src/Kore/Internal/SideCondition.hs

```diff
@@ -94,94 +97,115 @@ instance TopBottom (SideCondition variable) where
 instance InternalVariable variable
     => HasFreeVariables (SideCondition variable) variable
   where
-    freeVariables sideCondition@(SideCondition _) =
-        freeVariables assumedTrue
-      where
-        SideCondition {assumedTrue} = sideCondition
+    freeVariables = freeVariables . toPredicate
 
 instance InternalVariable variable => Unparse (SideCondition variable) where
-    unparse sideCondition@(SideCondition _) =
-        unparse assumedTrue
-      where
-        SideCondition {assumedTrue} = sideCondition
-
-    unparse2 sideCondition@(SideCondition _) =
-        unparse2 assumedTrue
-      where
-        SideCondition {assumedTrue} = sideCondition
+    unparse = unparse . toPredicate
+    unparse2 = unparse2 . toPredicate
 
-instance From (Condition variable) (SideCondition variable)
+instance From (SideCondition variable) (MultiAnd (Predicate variable))
   where
-      from = SideCondition
+    from condition@(SideCondition _) = assumedTrue condition
+    {-# INLINE from #-}
 
-instance From (SideCondition variable) (Condition variable) where
-    from = assumedTrue
+instance From (MultiAnd (Predicate variable)) (SideCondition variable)
+  where
+    from = SideCondition
     {-# INLINE from #-}
 
 instance
     InternalVariable variable
     => From (SideCondition variable) (Predicate variable)
   where
-    from = from @(Condition variable) . from @(SideCondition variable)
+    from = toPredicate
     {-# INLINE from #-}
 
 instance
     InternalVariable variable
     => From (Predicate variable) (SideCondition variable)
   where
-    from = from @(Condition variable) . from @(Predicate variable)
+    from = fromPredicate
+    {-# INLINE from #-}
+
+instance InternalVariable variable =>
+    From (Condition variable) (SideCondition variable)
+  where
+    from = fromCondition
+    {-# INLINE from #-}
+
+instance InternalVariable variable =>
+    From (SideCondition variable) (Condition variable)
+  where
+    from = Condition.fromPredicate . toPredicate
     {-# INLINE from #-}
 
-top :: InternalVariable variable => SideCondition variable
-top = fromCondition Condition.top
+top :: forall variable . SideCondition variable
+top =
+    from
+        @(MultiAnd (Predicate variable))
+        @(SideCondition variable)
+        MultiAnd.top
 
 -- | A 'top' 'Condition' for refactoring which should eventually be removed.
-topTODO :: InternalVariable variable => SideCondition variable
+topTODO :: SideCondition variable
 topTODO = top
 
 andCondition
     :: InternalVariable variable
     => SideCondition variable
     -> Condition variable
     -> SideCondition variable
-andCondition SideCondition { assumedTrue } newCondition =
+andCondition
+    SideCondition { assumedTrue }
+    (from @(Condition _) @(MultiAnd (Predicate _)) -> newPredicate)
+  =
     SideCondition merged
   where
-    merged = assumedTrue `Condition.andCondition` newCondition
+    merged = newPredicate <> assumedTrue
```

Comment:
If we can derive `Semigroup` above, I think this would be clearer:

```.hs
andCondition sideCondition newCondition =
    sideCondition <> fromCondition newCondition
```